harry-integration/test/harry/dsl/HistoryBuilderIntegrationTest.java

```diff
@@ -0,0 +1,141 @@
+                            return history.insert(entropySource.nextInt(maxPartitionSize));
+                        })
+                        .step((history) -> {
+                            int row = entropySource.nextInt(maxPartitionSize);
+                            long[] vds = new long[valueGenerators.length];
+                            for (int j = 0; j < valueGenerators.length; j++)
+                                vds[j] = valueGenerators[j].getAsLong();
+
+                            return history.insert(row, vds);
+                        })
+                        .step((history) -> {
+                            return history.deleteRow();
+                        })
+                        .step((history) -> {
+                            return history.deleteRow(entropySource.nextInt(maxPartitionSize));
+                        })
+                        .step(SingleOperationBuilder::deletePartition)
+                        .step(SingleOperationBuilder::deleteColumns)
+                        .step(SingleOperationBuilder::deleteRowSlice)
+                        .step((history) -> {
+                            return history.deleteRowRange();
+                        })
+                        .step((history) -> {
+                            return history.deleteRowRange(entropySource.nextInt(maxPartitionSize),
+                                                          entropySource.nextInt(maxPartitionSize),
+                                                          entropySource.nextBoolean(),
+                                                          entropySource.nextBoolean());
+                        })
+                        .step((history) -> history instanceof HistoryBuilder,
+                              (history) -> ((HistoryBuilder) history).beginBatch())
+                        .step((history) -> (history instanceof BatchVisitBuilder) && ((BatchVisitBuilder) history).size() > 1,
+                              (history) -> ((BatchVisitBuilder) history).endBatch())
+                        .exitCondition((history) -> {
+                            if (!(history instanceof HistoryBuilder))
+                                return false;
+
+                            HistoryBuilder historyBuilder = (HistoryBuilder) history;
+                            ReplayingVisitor visitor = historyBuilder.visitor(run.tracker, run.sut);
+                            visitor.replayAll();
+
+                            if (historyBuilder.presetSelector.pds().size() < MAX_PARTITIONS)
+                                return false;
+
+                            Model model = historyBuilder.checker(run.tracker, sut);
+
+                            for (Long pd : historyBuilder.presetSelector.pds())
+                                model.validate(Query.selectPartition(run.schemaSpec, pd,false));
+
+                            return true;
+                        })
+                        .run(STEPS_PER_ITERATION, seed);
```

Comment:
So I'll summarize what I think is going on here, and then my review should be complete...

We create a `ModelChecker` with single-operation steps to...
1.) Record a series of operations against the model using a `HistoryBuilder` as the state container.
2.) Replay them all against the SUT.
3.) Validate all the partitions we visited.

There are some details I'm not clear on, though. It looks like `ModelChecker#run()` executes individual steps, and each step execution adds data to the `HistoryBuilder` log. Then when we check the exit condition, we actually replay the operations to the SUT from the beginning up to the last visited LTS and validate the model.

Am I one the right track?

Reply:
Right, we create a model checker that triggers series of random steps, then replays them against the SUT, and validates visited partitions.

>  Then when we check the exit condition, we actually replay the operations to the SUT from the beginning up to the last visited LTS and validate the model.

So there are two different history builders: one replaying (with auto-replay on) and one that simply creates series of events, in case you would like to preturb / execute them in some sophisticated way. There was no specific way for us to collect and then replay things, but I have first created a regular history builder, and then replaying one, built on top of the first one.